client/src/components/layers/Layers.vue

```diff
@@ -0,0 +1,380 @@
+import Track, { TrackId } from '@/lib/track';
+import { FeaturePointingTarget } from '@/use/useFeaturePointing';
+import IntervalTree from '@flatten-js/interval-tree';
+
+
+import AnnotationLayer from '@/components/layers/AnnotationLayer';
+import TextLayer from '@/components/layers/TextLayer';
+import EditAnnotationLayer from '@/components/layers/EditAnnotationLayer.vue';
+import MarkerLayer from '@/components/layers/MarkerLayer.vue';
+import { geojsonToBound, geojsonToBound2 } from '../../utils';
+
+
+export default defineComponent({
+  props: {
+    trackMap: {
+      type: Map as PropType<Map<TrackId, Track>>,
+      required: true,
+    },
+    filteredTrackIds: {
+      type: Object as PropType<Ref<Array<TrackId>>>,
+      required: true,
+    },
+    intervalTree: {
+      type: Object as PropType<IntervalTree>,
+      required: true,
+    },
+    selectedTrackId: {
+      type: Object as PropType<Ref<TrackId>>,
+      required: true,
+    },
+    editingTrack: {
+      type: Object as PropType<Ref<boolean>>,
+      required: true,
+    },
+    typeColorMapper: {
+      type: Function as PropType<(t: string) => string>,
+      required: true,
+    },
+    stateStyling: {
+      type: Object,
+      required: true,
+    },
+    featurePointing: {
+      type: Object as PropType<Ref<boolean>>,
+      required: true,
+    },
+    featurePointingTarget: {
+      type: Object as PropType<Ref<FeaturePointingTarget>>,
+      required: true,
+    },
+  },
+
+  setup(props, { emit }) {
+    const annotator: any = inject('annotator');
+    const frameNumber: Readonly<Ref<number>> = computed(() => annotator.frame as number);
+
+
+    const Clicked = (trackId: number, editing: boolean) => {
+      emit('selectTrack', trackId, editing);
+    };
+
+    const annotationLayer = new AnnotationLayer({
+      annotator,
+      stateStyling: props.stateStyling,
+      typeColorMap: props.typeColorMapper,
+      signals: {
+        annotationClicked: Clicked,
+        annotationRightClicked: Clicked,
+      },
+    });
+
+    const textLayer = new TextLayer({
+      annotator,
+      stateStyling: props.stateStyling,
+      typeColorMap: props.typeColorMapper,
+    });
+
+    function updateLayers() {
+      // eslint-disable-next-line max-len
+      const currentFrameIds: number[] = props.intervalTree.search([frameNumber.value, frameNumber.value]);
+      const tracks: FrameDataTrack[] = [];
+      currentFrameIds.forEach(
+        (item: number) => {
+          if (props.filteredTrackIds.value.includes(item)) {
+            if (props.trackMap.has(item)) {
+              const track: Track = props.trackMap.get(item)!;
+              tracks.push({
+                selected: (props.selectedTrackId.value === track.trackId.value),
+                editing: props.editingTrack.value,
+                trackId: track.trackId.value,
+                features: track.features[frameNumber.value],
+                confidencePairs: track.confidencePairs.value,
```

Comment:
Why not replace `trackId, features, and confidencePairs` with a reference to the track itself?  This isn't a computed property, so it's safe to do.

Reply:
This keeps the `BaseLayers` from not knowing anything about the current frame.  If I pass the entire Track I'll also have to give them a frame number to know what slice of `features` to grab.

Reply:
By holding a reference to `annotator`, layers already have access to the current frame, you wouldn't need to provide it here.

Reply:
However, in the future, the process of getting frames will be more expensive than O(1), so I think doing it here is correct.  I'd recommend using `getFeature()` instead of accessing the feature array directly.

Reply:
This initial loop is also where I determine if the track is selected.  The true/false of it initially was mean to prevent Vue from doing stuff based on a numerical/string value changing.  It may make since for these (editingTrack, selectedTrackId) to just be passed to the BaseLayers in one spot instead of duplicating across all tracks in a frame.